lib/Model/ConvertFunctionType.cpp

```diff
@@ -0,0 +1,664 @@
+          return std::nullopt;
+        else if ((Index & 1) == 0 || ReturnStruct->Fields.size() <= 1
+                 || Index <= 1)
+          return std::nullopt;
+      }
+
+      MustUseTheNextOne = MustUseTheNextOne || IsCurrentRegisterUsed;
+    }
+
+    revng_assert(ReturnStruct->Size != 0 && !ReturnStruct->Fields.empty());
+    auto ReturnStructTypePath = TheBinary.recordNewType(std::move(Result));
+    revng_assert(ReturnStructTypePath.isValid());
+    return model::QualifiedType{ ReturnStructTypePath, {} };
+  }
+
+  return std::nullopt;
+}
+
+template<model::abi::Values V>
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionTypeImpl(const model::RawFunctionType &Function,
+                              model::Binary &TheBinary) {
+  using CC = ::abi::CallingConventionTrait<V>;
+  static constexpr model::Architecture::Values A = getArchitecture(V);
+  if (!verify<A>(Function.Arguments, CC::GeneralPurposeArgumentRegisters))
+    return std::nullopt;
+  if (!verify<A>(Function.ReturnValues, CC::GeneralPurposeReturnValueRegisters))
+    return std::nullopt;
+
+  constexpr model::Register::Values PTCRR = CC::ReturnValueLocationRegister;
+  if (PTCRR != model::Register::Invalid)
+    revng_assert(model::Register::getArchitecture(PTCRR) == A);
+  for (auto &SavedRegister : CC::CalleeSavedRegisters)
+    revng_assert(model::Register::getArchitecture(SavedRegister) == A);
+
+  model::CABIFunctionType Result;
+  Result.ABI = V;
+
+  auto ArgumentList = convertArguments<CC>(Function.Arguments,
+                                           CC::GeneralPurposeArgumentRegisters,
+                                           TheBinary);
+  if (ArgumentList == std::nullopt)
+    return std::nullopt;
+  for (auto &Argument : *ArgumentList)
+    Result.Arguments.insert(Argument);
+
+  auto
+    ReturnValue = convertReturnValue<CC>(Function.ReturnValues,
+                                         CC::GeneralPurposeReturnValueRegisters,
+                                         CC::ReturnValueLocationRegister,
+                                         TheBinary);
+  if (ReturnValue == std::nullopt)
+    return std::nullopt;
+  Result.ReturnType = *ReturnValue;
+  return Result;
+}
+
+std::optional<model::CABIFunctionType>
+convertToCABIFunctionType(const model::RawFunctionType &Function,
+                          model::abi::Values ABI,
+                          model::Binary &TheBinary) {
+  revng_assert(ABI != model::abi::Invalid);
+  return skippingEnumSwitch<1>(ABI, [&]<model::abi::Values A>() {
+    return convertToCABIFunctionTypeImpl<A>(Function, TheBinary);
+  });
+}
+
+//
+// toRaw
+//
+
+using IndexType = decltype(model::Argument::Index);
+using RegisterList = llvm::SmallVector<model::Register::Values, 1>;
+struct DistributedArgument {
+  RegisterList Registers = {};
+  size_t Size = 0, SizeOnStack = 0;
+};
+using DistributedArguments = llvm::DenseMap<IndexType, DistributedArgument>;
```

Comment:
Why a `DenseMap` instead of a simple `vector`?

Reply:
Wait, I hate being that guy who answers a question with a question but "why `SortedVector` instead of a simple `vector`?". I always assumed there's a deeper reason (maybe not every argument has to be present, etc. So I just kept it as close to the original as possible.

Reply:
That's due to limitations of `TupleTree`s.
We haven't implemented a proper diff strategy for "positional" containers (i.e., `vector`). We currently only diff containers that can be identified by a key, and vectors are not such containers (due to the fact the element of the vector is not aware of its position).

Once we'll have a diff strategy for "positional" containers, we'll promote the `SortedVector<Argument>` to a regular `vector`. But here you're not in the model, so you can go ahead and use `vector`.

Reply:
Thanks for the explanation, I'll keep that in mind.

Reply:
Done in https://github.com/revng/revng/pull/98/commits/a218c9dc571077f4f16d7aa6873d4d978ba57535.